packages/trpc/src/orderbook-router.ts

```diff
@@ -114,57 +122,75 @@ export const orderbookRouter = createTRPCRouter({
     }),
   getActiveOrders: publicProcedure
     .input(
-      z
-        .object({
-          contractOsmoAddress: z.string().startsWith("osmo"),
-          userOsmoAddress: z.string().startsWith("osmo"),
-        })
-        .required()
+      GetInfiniteLimitOrdersInputSchema.merge(
+        z.object({ contractOsmoAddress: z.string().startsWith("osmo") })
+      )
     )
     .query(async ({ input, ctx }) => {
-      const { contractOsmoAddress, userOsmoAddress } = input;
-      const resp = await getOrderbookActiveOrders({
-        orderbookAddress: contractOsmoAddress,
-        userOsmoAddress: userOsmoAddress,
-        chainList: ctx.chainList,
-      });
+      return maybeCachePaginatedItems({
+        getFreshItems: async () => {
+          const { contractOsmoAddress, userOsmoAddress } = input;
+          if (contractOsmoAddress.length === 0 || userOsmoAddress.length === 0)
+            return [];
+          const resp = await getOrderbookActiveOrders({
+            orderbookAddress: contractOsmoAddress,
+            userOsmoAddress: userOsmoAddress,
+            chainList: ctx.chainList,
+          });
 
-      if (resp.orders.length === 0) return [];
+          if (resp.orders.length === 0) return [];
 
-      return getTickInfoAndTransformOrders(
-        contractOsmoAddress,
-        resp.orders,
-        ctx.chainList
-      );
+          return getTickInfoAndTransformOrders(
+            contractOsmoAddress,
+            resp.orders,
+            ctx.chainList
+          );
+        },
+        cacheKey: JSON.stringify([
+          "active-orders",
+          input.contractOsmoAddress,
+          input.userOsmoAddress,
+        ]),
+      });
     }),
   getAllActiveOrders: publicProcedure
     .input(
-      z
-        .object({
-          contractAddresses: z.array(z.string().startsWith("osmo")),
-          userOsmoAddress: z.string().startsWith("osmo"),
-        })
-        .required()
+      GetInfiniteLimitOrdersInputSchema.merge(
+        z.object({ contractAddresses: z.array(z.string().startsWith("osmo")) })
+      )
     )
     .query(async ({ input, ctx }) => {
-      const { contractAddresses, userOsmoAddress } = input;
-      let allOrders: MappedLimitOrder[] = [];
-      for (let i = 0; i < contractAddresses.length; i++) {
-        const contractOsmoAddress = contractAddresses[i];
-        const resp = await getOrderbookActiveOrders({
-          orderbookAddress: contractOsmoAddress,
-          userOsmoAddress: userOsmoAddress,
-          chainList: ctx.chainList,
-        });
+      return maybeCachePaginatedItems({
+        getFreshItems: async () => {
+          const { contractAddresses, userOsmoAddress } = input;
+          if (contractAddresses.length === 0 || userOsmoAddress.length === 0)
+            return [];
+          let allOrders: MappedLimitOrder[] = [];
+          for (let i = 0; i < contractAddresses.length; i++) {
```

Comment:
`forEach` might be a little more idiomatic here

Reply:
I'm not sure `forEach` supports async operations does it? I could map it to an array of promises and run each loop asynchronously but not sure how that works with caching.

Reply:
https://github.com/osmosis-labs/osmosis-frontend/pull/3366/commits/e592dbbc08c38284b2545f103ad69791da36c753

Because the loop is using individual contracts per loop I can't see a caching conflict so I altered it to use promises instead (and `.map`)

Reply:
You are right that `forEach` will not wait for each block in the case of an async function. I like your map solution better 👍